### PR TITLE
Simplify AtomicBool::fetch_nand

### PR DESCRIPTION
**Pull request by mstewartgallus**
This should also be faster.

This code avoids storing in cases where there would not be a change.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Looks like tests are failing?

**Comment by nagisa** [comment]
I think you can simply `self.swap(true, order)` if `val` is false.

**Comment by sstewartgallus** [comment]
@nagis That is true. However, a swap implies a write all the time instead while a compare and swap does not which can be faster sometimes. In any case the bug was in the `if val` part so I replaced that with a `fetch_xor`. The commit will be up in a bit.

**Comment by alexcrichton** [comment]
The lack of a `loop` around the `false` branch here looks very worrisome to me, is there a reason why there's no loop around that?

**Comment by sstewartgallus** [comment]
@alexcrichton That's deliberate. !(x && false) = true. If x is true and not false then no write needs to happen which is good for the cache.

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -512,17 +512,26 @@ impl AtomicBool {
...
-                Err(x) => old = x,
+        // So we simplify to fetch_xor or a swap.
+
+        // !(true && true) == false
+        // !(true && false) == true
+        // !(false && false) == true
+        // !(false && true) == true
+
+        // !(x && true) == !x
+        // !(x && false) == true
```
Comment (eddyb) [inline src/libcore/sync/atomic.rs]: I would use `&` instead of `&&` in these, seems clearer.

File: src/libcore/sync/atomic.rs [hunk 2]
```diff
@@ -512,17 +512,26 @@ impl AtomicBool {
...
+        if val {
+            return self.fetch_xor(true, order);
```
Comment (eddyb) [inline src/libcore/sync/atomic.rs]: Explicit `return` isn't needed.

**Comment by sstewartgallus** [comment]
This is odd https://travis-ci.org/rust-lang/rust/jobs/213176849 I can't view what happened here.

**Comment by nagisa** [comment]
To me this all seems correct. I still think atomoc_swap would be better in
a sense that it's easier to see it's obviously correct unlike he CAS but
either way seems correct to me.

**Comment by eddyb** [comment]
@sstewartgallus Only one configuration is supposed to run on a PR, but the only way to do that in travis is to finish immediately for non-PR configurations. But macOS ones sometimes even fail at that.

**Comment by alexcrichton** [comment]
Do you have benchmarks for this as well to justify this change?

**Comment by sstewartgallus** [comment]
@alexcrichton Fair enough. However, as this is in libcore I can't easily use std::threads for benchmarking purposes can I?

**Comment by eddyb** [comment]
@sstewartgallus You can write your own standalone benchmark.

**Comment by stjepang** [comment]
Benchmark: [source](https://gist.github.com/stjepang/b0dc898c12467186c5796de74fdf395c)
Results:

```
test new1_1t ... bench:      89,684 ns/iter (+/- 13,753)
test new1_2t ... bench:     315,369 ns/iter (+/- 376,217)
test new1_4t ... bench:   1,029,476 ns/iter (+/- 433,656)
test new2_1t ... bench:      89,585 ns/iter (+/- 17,570)
test new2_2t ... bench:     336,093 ns/iter (+/- 375,732)
test new2_4t ... bench:     994,311 ns/iter (+/- 312,345)
test old_1t  ... bench:     147,787 ns/iter (+/- 23,592)
test old_2t  ... bench:     656,805 ns/iter (+/- 796,442)
test old_4t  ... bench:   2,869,017 ns/iter (+/- 1,074,877)
```

Prety good imprevements! :)

Let's just get rid of that unnecessary `return` and add a test for `fetch_nand` to `src/libcore/tests/atomic.rs`.

**Comment by stjepang** [comment]
Let's close this PR.

**Pull request closed without merging**